File: ai-engine/services/audit_service.py
from datetime import datetime


class AuditService:

    def __init__(self):

        self.logs = []

    def log(
        self,
        action,
        case_id=None
    ):

        entry = {

            "timestamp":
                datetime.now()
                .strftime(
                    "%Y-%m-%d %H:%M:%S"
                ),

            "caseId":
                case_id,

            "action":
                action
        }

        self.logs.append(
            entry
        )

        return entry

    def get_logs(self):

        return self.logs

File: backend/src/audit/audit.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';

@Injectable()
export class AuditService {
  constructor(
    private prisma: PrismaService,
  ) {}

  async log(
    action: string,
    userId: string,
    caseId?: string,
    details?: string,
  ) {
    return this.prisma.auditLog.create({
      data: {
        action,
        userId,
        caseId,
        details,
      },
    });
  }

  async findAll() {
    return this.prisma.auditLog.findMany({
      include: {
        user: true,
      },
      orderBy: {
        createdAt: 'desc',
      },
    });
  }

 async findCaseTimeline(caseId: string) {

  const logs =
    await this.prisma.auditLog.findMany({
      where: {
        caseId,
      },
      include: {
        user: true,
      },
      orderBy: {
        createdAt: 'asc',
      },
    });

  return logs.map((log) => ({
    timestamp: log.createdAt,

    officer:
      log.user.name,

    event:
      log.action.replaceAll('_', ' '),

    details:
      log.details,
  }));
}
}